cds-snc/cppd-medical-report-sails
# Testing the patient flow

# What this does
Adds a whole bunch of cypress tests for the Patient flow. Including:
- Tests validation on the Personal form
    - Tests valid/invalid dates (these might be better covered in a unit test?)
    - Tests valid/invalid SINs (these are also covered in unit tests, so maybe redundant?)
- Tests validation on the Consent form
- Tests the signature pad/text input toggle on the Consent form
- Unit tests the SIN validator
- Tests email validation (if an email is provided)
- Tests the happy path for the Patient flow
- Introduces a `social-insurance-number` package from Wealthsimple that enables generating and validating SINs - useful for testing (and will incorporate in our SIN validator soon)
- Introduces the `faker` library for generating fake data for tests

Pretty new at this type of testing... there may be gaps, there may be redundancies, but it is a start!

## What it doesn't test
- I don't know how to test drawing on the signature pad on the consent form yet.

## Test file organization
Cypress doesn't provide a lot of guidance on how to organize test files as every application is different. I'll describe here my thinking in how I've grouped these, but this may evolve over time, and open to suggestions.

`patient_flow.spec.js` - this spec covers the overall flow from the Patient's perspective. This is a happy-path test, just ensuring that the patient can make it through the forms as expected with normal input.

`personal.spec.js` and `consent.spec.js` - these specs cover the individual forms that make up the patient flow. These perform detailed form validation testing and test various possible states of the forms.

`unit/validators/sin.spec.js` - cypress is normally used for end-to-end tests, not necessarily unit tests, but that doesn't mean you can't unit test. This spec is mainly a proof-of-concept and is a bit redundant since we test the sin validation on the personal form spec, but it works!

## To test
- run `npm install` to make sure you have the latest everything
- make sure you have the service running locally (`npm run dev`)
- run `npm run cypress:start` in another console to get cypress running
- select `Run all specs` in the cypress interface
